src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -1,16 +1,935 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
+using System;
+using System.Collections.Immutable;
 using System.Linq;
+using System.Text;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
 using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
 using Xunit;
-
 namespace Microsoft.CodeAnalysis.CSharp.UnitTests.CodeGen
 {
+    using static Instruction;
+    internal enum Instruction
+    {
+        Write,
+        Yield,
+        AwaitSlow,
+        AwaitFast,
+        YieldBreak
+    }
+
     [CompilerTrait(CompilerFeature.AsyncStreams)]
     public class CodeGenAsyncIteratorTests : EmitMetadataTestBase
     {
+        // PROTOTYPE(async-streams)
+        // Test missing types/members
+        // Test with yield or await in try/catch/finally
+        // More tests with exception thrown
+        // Test yield break
+        // There is a case in GetIteratorElementType with IsDirectlyInIterator that relates to speculation, needs testing
+        // yield break disallowed in finally and top-level script (see BindYieldBreakStatement); same for yield return (see BindYieldReturnStatement)
+        // test local function
+        // binding for yield return (BindYieldReturnStatement) validates escape rules, needs testing
+        // test yield in async lambda (still error)
+        // test exception handling (should capture and return the exception via the promise)
+        // test IAsyncEnumerable<U> M<U>() ...
+        // test missing AsyncTaskMethodBuilder<T> or missing members Create(), Task, ...
+        // test with IAsyncEnumerable<dynamic>
+        // other tests with dynamic?
+        // test should cover both case with AwaitOnCompleted and AwaitUnsafeOnCompleted
+        // test `async IAsyncEnumerable<int> M() { return TaskLike(); }`
+        // Can we avoid making IAsyncEnumerable<T> special from the start? Making mark it with an attribute like we did for task-like?
+        // Test normal break in async-iterator
+        // WaitForNextAsync is resilient to be called out of turn. Test that.
+        // Test a plain return statement
+
+        [Fact]
+        public void AsyncIteratorWithAwaitCompletedAndYield()
```

Comment:
>AsyncIteratorWithAwaitCompletedAndYield [](start = 20, length = 39)

Maybe another example with release output to make sure any optimization changes are expected.